File: ui/src/utils/instructions.ts
import {
  PublicKey,
  TransactionInstruction,
  SystemProgram,
} from "@solana/web3.js";
import { Buffer } from "buffer";
import * as borsh from "borsh";
import {
  findFacilitatorRegistryPda,
  findBoardPda,
  findNotePda,
  findGroupPda,
  findBoardMembershipPda,
  findVoteRecordPda,
  findActionItemPda,
  findVerificationVotePda,
  findSessionTokenPda,
} from "./pda";

// Instruction discriminators (must match instructions.rs)
const INIT_FACILITATOR_REGISTRY = 0;
const CREATE_BOARD = 1;
const ADVANCE_STAGE = 2;
const CLOSE_BOARD = 3;
const CREATE_NOTE = 4;
const CREATE_GROUP = 5;
const SET_GROUP_TITLE = 6;
const ASSIGN_NOTE_TO_GROUP = 7;
const UNASSIGN_NOTE = 8;
const CAST_VOTE = 9;
const CREATE_ACTION_ITEM = 10;
const CAST_VERIFICATION_VOTE = 11;
const CREATE_SESSION = 12;
const REVOKE_SESSION = 13;

// Borsh schema definitions
const createBoardSchema = {
  struct: {
    categories: { array: { type: "string" } },
    allowlist: { array: { type: { array: { type: "u8", len: 32 } } } },
    voting_credits_per_participant: { option: "u8" },
  },
};

const advanceStageSchema = {
  struct: {
    new_stage: "u8",
  },
};

const createNoteSchema = {
  struct: {
    category_id: "u8",
    content: "string",
  },
};

const createGroupSchema = {
  struct: {
    title: "string",
  },
};

const setGroupTitleSchema = {
  struct: {
    group_id: "u64",
    title: "string",
  },
};

const assignNoteSchema = {
  struct: {
    note_id: "u64",
    group_id: "u64",
  },
};

const unassignNoteSchema = {
  struct: {
    note_id: "u64",
  },
};

const castVoteSchema = {
  struct: {
    group_id: "u64",
    credits_delta: "u8",
  },
};

const createActionItemSchema = {
  struct: {
    description: "string",
    owner: { array: { type: "u8", len: 32 } },
    verifiers: { array: { type: { array: { type: "u8", len: 32 } } } },
    threshold: "u8",
  },
};

const castVerificationVoteSchema = {
  struct: {
    action_item_id: "u64",
    approved: "bool",
  },
};

const createSessionSchema = {
  struct: {
    valid_until: "i64",
    top_up_lamports: { option: "u64" },
  },
};

function serializeInstruction(discriminator: number, payload?: Buffer): Buffer {
  if (payload) {
    return Buffer.concat([Buffer.from([discriminator]), payload]);
  }
  return Buffer.from([discriminator]);
}

export function createInitFacilitatorRegistryInstruction(
  facilitator: PublicKey,
  programId: PublicKey
): TransactionInstruction {
  const [facilitatorRegistry] = findFacilitatorRegistryPda(facilitator, programId);

  return new TransactionInstruction({
    keys: [
      { pubkey: facilitatorRegistry, isSigner: false, isWritable: true },
      { pubkey: facilitator, isSigner: true, isWritable: true },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
    ],
    programId,
    data: serializeInstruction(INIT_FACILITATOR_REGISTRY),
  });
}

export function createCreateBoardInstruction(
  facilitatorRegistry: PublicKey,
  board: PublicKey,
  signer: PublicKey,
  categories: string[],
  allowlist: PublicKey[],
  votingCreditsPerParticipant: number | null,
  membershipAccounts: PublicKey[],
  programId: PublicKey,
  sessionToken?: PublicKey
): TransactionInstruction {
  const payload = {
    categories,
    allowlist: allowlist.map((pk) => Array.from(pk.toBytes())),
    voting_credits_per_participant: votingCreditsPerParticipant,
  };

  const serialized = borsh.serialize(createBoardSchema as any, payload);

  const keys = [
    { pubkey: facilitatorRegistry, isSigner: false, isWritable: true },
    { pubkey: board, isSigner: false, isWritable: true },
    { pubkey: signer, isSigner: true, isWritable: true },
    { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
  ];

  // Session token comes after system_program but before membership accounts
  if (sessionToken) {
    keys.push({ pubkey: sessionToken, isSigner: false, isWritable: false });
  }

  // BoardMembership accounts for each allowlist member (enables board discovery)
  for (const ma of membershipAccounts) {
    keys.push({ pubkey: ma, isSigner: false, isWritable: true });
  }

  return new TransactionInstruction({
    keys,
    programId,
    data: serializeInstruction(CREATE_BOARD, Buffer.from(serialized)),
  });
}

export function createAdvanceStageInstruction(
  board: PublicKey,
  signer: PublicKey,
  newStage: number,
  programId: PublicKey,
  sessionToken?: PublicKey
): TransactionInstruction {
  const payload = { new_stage: newStage };
  const serialized = borsh.serialize(advanceStageSchema as any, payload);

  const keys = [
    { pubkey: board, isSigner: false, isWritable: true },
    { pubkey: signer, isSigner: true, isWritable: false },
  ];

  if (sessionToken) {
    keys.push({ pubkey: sessionToken, isSigner: false, isWritable: false });
  }

  return new TransactionInstruction({
    keys,
    programId,
    data: serializeInstruction(ADVANCE_STAGE, Buffer.from(serialized)),
  });
}

export function createCloseBoardInstruction(
  board: PublicKey,
  signer: PublicKey,
  programId: PublicKey,
  sessionToken?: PublicKey
): TransactionInstruction {
  const keys = [
    { pubkey: board, isSigner: false, isWritable: true },
    { pubkey: signer, isSigner: true, isWritable: false },
  ];

  if (sessionToken) {
    keys.push({ pubkey: sessionToken, isSigner: false, isWritable: false });
  }

  return new TransactionInstruction({
    keys,
    programId,
    data: serializeInstruction(CLOSE_BOARD),
  });
}

export function createCreateNoteInstruction(
  board: PublicKey,
  note: PublicKey,
  signer: PublicKey,
  categoryId: number,
  content: string,
  programId: PublicKey,
  sessionToken?: PublicKey
): TransactionInstruction {
  const payload = { category_id: categoryId, content };
  const serialized = borsh.serialize(createNoteSchema as any, payload);

  const keys = [
    { pubkey: board, isSigner: false, isWritable: true },
    { pubkey: note, isSigner: false, isWritable: true },
    { pubkey: signer, isSigner: true, isWritable: true },
    { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
  ];

  if (sessionToken) {
    keys.push({ pubkey: sessionToken, isSigner: false, isWritable: false });
  }

  return new TransactionInstruction({
    keys,
    programId,
    data: serializeInstruction(CREATE_NOTE, Buffer.from(serialized)),
  });
}

export function createCreateGroupInstruction(
  board: PublicKey,
  group: PublicKey,
  signer: PublicKey,
  title: string,
  programId: PublicKey,
  sessionToken?: PublicKey
): TransactionInstruction {
  const payload = { title };
  const serialized = borsh.serialize(createGroupSchema as any, payload);

  const keys = [
    { pubkey: board, isSigner: false, isWritable: true },
    { pubkey: group, isSigner: false, isWritable: true },
    { pubkey: signer, isSigner: true, isWritable: true },
    { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
  ];

  if (sessionToken) {
    keys.push({ pubkey: sessionToken, isSigner: false, isWritable: false });
  }

  return new TransactionInstruction({
    keys,
    programId,
    data: serializeInstruction(CREATE_GROUP, Buffer.from(serialized)),
  });
}

export function createSetGroupTitleInstruction(
  board: PublicKey,
  group: PublicKey,
  signer: PublicKey,
  groupId: bigint,
  title: string,
  programId: PublicKey,
  sessionToken?: PublicKey
): TransactionInstruction {
  const payload = { group_id: groupId, title };
  const serialized = borsh.serialize(setGroupTitleSchema as any, payload);

  const keys = [
    { pubkey: board, isSigner: false, isWritable: false },
    { pubkey: group, isSigner: false, isWritable: true },
    { pubkey: signer, isSigner: true, isWritable: false },
  ];

  if (sessionToken) {
    keys.push({ pubkey: sessionToken, isSigner: false, isWritable: false });
  }

  return new TransactionInstruction({
    keys,
    programId,
    data: serializeInstruction(SET_GROUP_TITLE, Buffer.from(serialized)),
  });
}

export function createAssignNoteToGroupInstruction(
  board: PublicKey,
  note: PublicKey,
  group: PublicKey,
  signer: PublicKey,
  noteId: bigint,
  groupId: bigint,
  programId: PublicKey,
  sessionToken?: PublicKey
): TransactionInstruction {
  const payload = { note_id: noteId, group_id: groupId };
  const serialized = borsh.serialize(assignNoteSchema as any, payload);

  const keys = [
    { pubkey: board, isSigner: false, isWritable: false },
    { pubkey: note, isSigner: false, isWritable: true },
    { pubkey: group, isSigner: false, isWritable: false },
    { pubkey: signer, isSigner: true, isWritable: false },
  ];

  if (sessionToken) {
    keys.push({ pubkey: sessionToken, isSigner: false, isWritable: false });
  }

  return new TransactionInstruction({
    keys,
    programId,
    data: serializeInstruction(ASSIGN_NOTE_TO_GROUP, Buffer.from(serialized)),
  });
}

export function createUnassignNoteInstruction(
  board: PublicKey,
  note: PublicKey,
  signer: PublicKey,
  noteId: bigint,
  programId: PublicKey,
  sessionToken?: PublicKey
): TransactionInstruction {
  const payload = { note_id: noteId };
  const serialized = borsh.serialize(unassignNoteSchema as any, payload);

  const keys = [
    { pubkey: board, isSigner: false, isWritable: false },
    { pubkey: note, isSigner: false, isWritable: true },
    { pubkey: signer, isSigner: true, isWritable: false },
  ];

  if (sessionToken) {
    keys.push({ pubkey: sessionToken, isSigner: false, isWritable: false });
  }

  return new TransactionInstruction({
    keys,
    programId,
    data: serializeInstruction(UNASSIGN_NOTE, Buffer.from(serialized)),
  });
}

export function createCastVoteInstruction(
  board: PublicKey,
  boardMembership: PublicKey,
  group: PublicKey,
  voteRecord: PublicKey,
  signer: PublicKey,
  groupId: bigint,
  creditsDelta: number,
  programId: PublicKey,
  sessionToken?: PublicKey
): TransactionInstruction {
  const payload = { group_id: groupId, credits_delta: creditsDelta };
  const serialized = borsh.serialize(castVoteSchema as any, payload);

  const keys = [
    { pubkey: board, isSigner: false, isWritable: false },
    { pubkey: boardMembership, isSigner: false, isWritable: true },
    { pubkey: group, isSigner: false, isWritable: true },
    { pubkey: voteRecord, isSigner: false, isWritable: true },
    { pubkey: signer, isSigner: true, isWritable: true },
    { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
  ];

  if (sessionToken) {
    keys.push({ pubkey: sessionToken, isSigner: false, isWritable: false });
  }

  return new TransactionInstruction({
    keys,
    programId,
    data: serializeInstruction(CAST_VOTE, Buffer.from(serialized)),
  });
}

export function createCreateActionItemInstruction(
  board: PublicKey,
  actionItem: PublicKey,
  signer: PublicKey,
  description: string,
  owner: PublicKey,
  verifiers: PublicKey[],
  threshold: number,
  programId: PublicKey,
  sessionToken?: PublicKey
): TransactionInstruction {
  const payload = {
    description,
    owner: Array.from(owner.toBytes()),
    verifiers: verifiers.map((pk) => Array.from(pk.toBytes())),
    threshold,
  };
  const serialized = borsh.serialize(createActionItemSchema as any, payload);

  const keys = [
    { pubkey: board, isSigner: false, isWritable: true },
    { pubkey: actionItem, isSigner: false, isWritable: true },
    { pubkey: signer, isSigner: true, isWritable: true },
    { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
  ];

  if (sessionToken) {
    keys.push({ pubkey: sessionToken, isSigner: false, isWritable: false });
  }

  return new TransactionInstruction({
    keys,
    programId,
    data: serializeInstruction(CREATE_ACTION_ITEM, Buffer.from(serialized)),
  });
}

export function createCastVerificationVoteInstruction(
  board: PublicKey,
  actionItem: PublicKey,
  verificationVote: PublicKey,
  ownerMembership: PublicKey,
  signer: PublicKey,
  actionItemId: bigint,
  approved: boolean,
  programId: PublicKey,
  sessionToken?: PublicKey
): TransactionInstruction {
  const payload = { action_item_id: actionItemId, approved };
  const serialized = borsh.serialize(castVerificationVoteSchema as any, payload);

  const keys = [
    { pubkey: board, isSigner: false, isWritable: false },
    { pubkey: actionItem, isSigner: false, isWritable: true },
    { pubkey: verificationVote, isSigner: false, isWritable: true },
    { pubkey: ownerMembership, isSigner: false, isWritable: true },
    { pubkey: signer, isSigner: true, isWritable: true },
    { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
  ];

  if (sessionToken) {
    keys.push({ pubkey: sessionToken, isSigner: false, isWritable: false });
  }

  return new TransactionInstruction({
    keys,
    programId,
    data: serializeInstruction(CAST_VERIFICATION_VOTE, Buffer.from(serialized)),
  });
}

// Session key instructions

export function createCreateSessionInstruction(
  sessionToken: PublicKey,
  sessionSigner: PublicKey,
  authority: PublicKey,
  validUntil: bigint,
  topUpLamports: bigint | null,
  programId: PublicKey
): TransactionInstruction {
  const payload = {
    valid_until: validUntil,
    top_up_lamports: topUpLamports,
  };
  const serialized = borsh.serialize(createSessionSchema as any, payload);

  return new TransactionInstruction({
    keys: [
      { pubkey: sessionToken, isSigner: false, isWritable: true },
      { pubkey: sessionSigner, isSigner: true, isWritable: true },
      { pubkey: authority, isSigner: true, isWritable: true },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
    ],
    programId,
    data: serializeInstruction(CREATE_SESSION, Buffer.from(serialized)),
  });
}

export function createRevokeSessionInstruction(
  sessionToken: PublicKey,
  authority: PublicKey,
  programId: PublicKey
): TransactionInstruction {
  return new TransactionInstruction({
    keys: [
      { pubkey: sessionToken, isSigner: false, isWritable: true },
      { pubkey: authority, isSigner: true, isWritable: true },
    ],
    programId,
    data: serializeInstruction(REVOKE_SESSION),
  });
}

// Re-export PDA helpers for convenience
export {
  findFacilitatorRegistryPda,
  findBoardPda,
  findNotePda,
  findGroupPda,
  findBoardMembershipPda,
  findVoteRecordPda,
  findActionItemPda,
  findVerificationVotePda,
  findSessionTokenPda,
};

File: programs/retroquest/src/instructions.rs
use borsh::BorshDeserialize;
use solana_program::{program_error::ProgramError, pubkey::Pubkey};

use crate::state::BoardStage;

// Instruction discriminators
pub const INIT_FACILITATOR_REGISTRY: u8 = 0;
pub const CREATE_BOARD: u8 = 1;
pub const ADVANCE_STAGE: u8 = 2;
pub const CLOSE_BOARD: u8 = 3;
pub const CREATE_NOTE: u8 = 4;
pub const CREATE_GROUP: u8 = 5;
pub const SET_GROUP_TITLE: u8 = 6;
pub const ASSIGN_NOTE_TO_GROUP: u8 = 7;
pub const UNASSIGN_NOTE: u8 = 8;
pub const CAST_VOTE: u8 = 9;
pub const CREATE_ACTION_ITEM: u8 = 10;
pub const CAST_VERIFICATION_VOTE: u8 = 11;

#[derive(Debug)]
pub enum RetroInstruction {
    /// Initialize a facilitator registry
    /// Accounts:
    /// 0. `[writable]` Facilitator registry PDA
    /// 1. `[signer]` Facilitator
    /// 2. `[]` System program
    InitFacilitatorRegistry,

    /// Create a new retro board
    /// Accounts:
    /// 0. `[writable]` Facilitator registry PDA
    /// 1. `[writable]` Board PDA
    /// 2. `[signer]` Facilitator
    /// 3. `[]` System program
    /// 4+ `[writable]` BoardMembership PDAs for each allowlist member
    CreateBoard {
        categories: Vec<String>,
        allowlist: Vec<Pubkey>,
        voting_credits_per_participant: Option<u8>,
    },

    /// Advance board to next stage
    /// Accounts:
    /// 0. `[writable]` Board PDA
    /// 1. `[signer]` Facilitator
    AdvanceStage { new_stage: BoardStage },

    /// Close the board
    /// Accounts:
    /// 0. `[writable]` Board PDA
    /// 1. `[signer]` Facilitator
    CloseBoard,

    /// Create a note (must be on allowlist)
    /// Accounts:
    /// 0. `[writable]` Board PDA
    /// 1. `[writable]` Note PDA
    /// 2. `[signer]` Author
    /// 3. `[]` System program
    CreateNote { category_id: u8, content: String },

    /// Create a group (must be on allowlist)
    /// Accounts:
    /// 0. `[writable]` Board PDA
    /// 1. `[writable]` Group PDA
    /// 2. `[signer]` Creator
    /// 3. `[]` System program
    CreateGroup { title: String },

    /// Set group title (must be on allowlist)
    /// Accounts:
    /// 0. `[]` Board PDA
    /// 1. `[writable]` Group PDA
    /// 2. `[signer]` Participant
    SetGroupTitle { group_id: u64, title: String },

    /// Assign note to group (must be on allowlist)
    /// Accounts:
    /// 0. `[]` Board PDA
    /// 1. `[writable]` Note PDA
    /// 2. `[]` Group PDA
    /// 3. `[signer]` Participant
    AssignNoteToGroup { note_id: u64, group_id: u64 },

    /// Unassign note from group (must be on allowlist)
    /// Accounts:
    /// 0. `[]` Board PDA
    /// 1. `[writable]` Note PDA
    /// 2. `[signer]` Participant
    UnassignNote { note_id: u64 },

    /// Cast vote (must be on allowlist)
    /// Uses BoardMembership to track credits
    /// Accounts:
    /// 0. `[]` Board PDA
    /// 1. `[writable]` BoardMembership PDA
    /// 2. `[writable]` Group PDA
    /// 3. `[writable]` Vote record PDA
    /// 4. `[signer]` Voter
    /// 5. `[]` System program
    CastVote { group_id: u64, credits_delta: u8 },

    /// Create an action item (facilitator only, Discuss stage)
    /// Accounts:
    /// 0. `[writable]` Board PDA
    /// 1. `[writable]` ActionItem PDA
    /// 2. `[signer]` Facilitator
    /// 3. `[]` System program
    CreateActionItem {
        description: String,
        owner: Pubkey,
        verifiers: Vec<Pubkey>,
        threshold: u8,
    },

    /// Cast a verification vote on an action item (board must be closed)
    /// Accounts:
    /// 0. `[]` Board PDA
    /// 1. `[writable]` ActionItem PDA
    /// 2. `[writable]` VerificationVote PDA
    /// 3. `[writable]` Owner's BoardMembership PDA (for score update)
    /// 4. `[signer]` Verifier
    /// 5. `[]` System program
    CastVerificationVote { action_item_id: u64, approved: bool },
}

// Instruction data payloads for Borsh deserialization
#[derive(BorshDeserialize)]
struct CreateBoardPayload {
    categories: Vec<String>,
    allowlist: Vec<Pubkey>,
    voting_credits_per_participant: Option<u8>,
}

#[derive(BorshDeserialize)]
struct AdvanceStagePayload {
    new_stage: u8,
}

#[derive(BorshDeserialize)]
struct CreateNotePayload {
    category_id: u8,
    content: String,
}

#[derive(BorshDeserialize)]
struct CreateGroupPayload {
    title: String,
}

#[derive(BorshDeserialize)]
struct SetGroupTitlePayload {
    group_id: u64,
    title: String,
}

#[derive(BorshDeserialize)]
struct AssignNotePayload {
    note_id: u64,
    group_id: u64,
}

#[derive(BorshDeserialize)]
struct UnassignNotePayload {
    note_id: u64,
}

#[derive(BorshDeserialize)]
struct CastVotePayload {
    group_id: u64,
    credits_delta: u8,
}

#[derive(BorshDeserialize)]
struct CreateActionItemPayload {
    description: String,
    owner: Pubkey,
    verifiers: Vec<Pubkey>,
    threshold: u8,
}

#[derive(BorshDeserialize)]
struct CastVerificationVotePayload {
    action_item_id: u64,
    approved: bool,
}

impl RetroInstruction {
    pub fn unpack(input: &[u8]) -> Result<Self, ProgramError> {
        let (&variant, rest) = input
            .split_first()
            .ok_or(ProgramError::InvalidInstructionData)?;

        Ok(match variant {
            0 => Self::InitFacilitatorRegistry,

            1 => {
                let payload = CreateBoardPayload::try_from_slice(rest)
                    .map_err(|_| ProgramError::InvalidInstructionData)?;
                Self::CreateBoard {
                    categories: payload.categories,
                    allowlist: payload.allowlist,
                    voting_credits_per_participant: payload.voting_credits_per_participant,
                }
            }

            2 => {
                let payload = AdvanceStagePayload::try_from_slice(rest)
                    .map_err(|_| ProgramError::InvalidInstructionData)?;
                let new_stage = match payload.new_stage {
                    0 => BoardStage::Setup,
                    1 => BoardStage::WriteNotes,
                    2 => BoardStage::GroupDuplicates,
                    3 => BoardStage::Vote,
                    4 => BoardStage::Discuss,
                    _ => return Err(ProgramError::InvalidInstructionData),
                };
                Self::AdvanceStage { new_stage }
            }

            3 => Self::CloseBoard,

            4 => {
                let payload = CreateNotePayload::try_from_slice(rest)
                    .map_err(|_| ProgramError::InvalidInstructionData)?;
                Self::CreateNote {
                    category_id: payload.category_id,
                    content: payload.content,
                }
            }

            5 => {
                let payload = CreateGroupPayload::try_from_slice(rest)
                    .map_err(|_| ProgramError::InvalidInstructionData)?;
                Self::CreateGroup {
                    title: payload.title,
                }
            }

            6 => {
                let payload = SetGroupTitlePayload::try_from_slice(rest)
                    .map_err(|_| ProgramError::InvalidInstructionData)?;
                Self::SetGroupTitle {
                    group_id: payload.group_id,
                    title: payload.title,
                }
            }

            7 => {
                let payload = AssignNotePayload::try_from_slice(rest)
                    .map_err(|_| ProgramError::InvalidInstructionData)?;
                Self::AssignNoteToGroup {
                    note_id: payload.note_id,
                    group_id: payload.group_id,
                }
            }

            8 => {
                let payload = UnassignNotePayload::try_from_slice(rest)
                    .map_err(|_| ProgramError::InvalidInstructionData)?;
                Self::UnassignNote {
                    note_id: payload.note_id,
                }
            }

            9 => {
                let payload = CastVotePayload::try_from_slice(rest)
                    .map_err(|_| ProgramError::InvalidInstructionData)?;
                Self::CastVote {
                    group_id: payload.group_id,
                    credits_delta: payload.credits_delta,
                }
            }

            10 => {
                let payload = CreateActionItemPayload::try_from_slice(rest)
                    .map_err(|_| ProgramError::InvalidInstructionData)?;
                Self::CreateActionItem {
                    description: payload.description,
                    owner: payload.owner,
                    verifiers: payload.verifiers,
                    threshold: payload.threshold,
                }
            }

            11 => {
                let payload = CastVerificationVotePayload::try_from_slice(rest)
                    .map_err(|_| ProgramError::InvalidInstructionData)?;
                Self::CastVerificationVote {
                    action_item_id: payload.action_item_id,
                    approved: payload.approved,
                }
            }

            _ => return Err(ProgramError::InvalidInstructionData),
        })
    }
}
